src/docs/arch-cursor.dox

```diff
@@ -3,13 +3,393 @@
 Cursors are used in WiredTiger to get and modify data.
 A caller of WiredTiger uses WT_SESSION::open_cursor to create
 a WT_CURSOR.  Methods on the WT_CURSOR can then be used to
-position, iterate, get, and set data.
+position, iterate, get, and set data.  In the typical case, a cursor
+will be used to access keys and values in a Btree.  Cursors are
+also used to retrieve indexed data as well as WiredTiger statistics, log files,
+and metadata. Additionally, cursors are used for managing backups.
 
-Depending on the <code>uri</code> used when creating a cursor, the cursor will
-be internally implemented as one of the many cursor structures that include
-WT_CURSOR_BTREE, WT_CURSOR_BACKUP, WT_CURSOR_INDEX, WT_CURSOR_LOG,
-WT_CURSOR_METADATA, WT_CURSOR_STAT. Each of these structures starts
+The various kinds of cursors are created during the WT_SESSION::open_cursor call.
+Depending on the <code>uri</code> used when opening a cursor, the cursor will
+be implemented internally as one of the many cursor structures that include
+\c WT_CURSOR_BTREE, \c WT_CURSOR_BACKUP, \c WT_CURSOR_INDEX, \c WT_CURSOR_LOG,
+\c WT_CURSOR_METADATA, \c WT_CURSOR_STAT, \c WT_CURSOR_TABLE. Each of these structures starts
 with the common \c %WT_CURSOR structure, which contain all
 of the data and method pointers that make up the public part of the API.
+Thus, any one of these "extended cursor structs" can be allocated and
+returned as a WT_CURSOR pointer.  Since the method pointers are filled with
+a specific implementations, for example \c __curtable_reset, a call to
+WT_CURSOR::reset will call the specific implementation function, and the first
+argument can be cast to be a pointer to the cursor struct used by the implementation.
+Thus, in our C code, we have something similar to a class hierarchy, having
+an abstract base class (WT_CURSOR) with virtual methods, and a number of
+implementation classes.
+
+The code for each cursor type's methods are generally grouped together,
```

Comment:
The code for each cursor type's methods is generally organized in a single file.

Reply:
done.